storage/src/merkle_tree_db_adapter.cpp

```diff
@@ -0,0 +1,323 @@
+// Copyright 2020 VMware, all rights reserved
+
+#include "blockchain/db_types.h"
+#include "blockchain/merkle_tree_db_adapter.h"
+#include "Logger.hpp"
+#include "sparse_merkle/base_types.h"
+#include "sliver.hpp"
+#include "status.hpp"
+#include "endianness.hpp"
+
+#include <exception>
+#include <iterator>
+#include <limits>
+#include <string>
+#include <type_traits>
+#include <vector>
+#include <utility>
+
+using concordUtils::Sliver;
+using concordUtils::Status;
+
+using concord::storage::blockchain::v2MerkleTree::detail::EDBKeyType;
+using concord::storage::blockchain::v2MerkleTree::detail::EKeyType;
+using concord::storage::blockchain::v2MerkleTree::detail::EBFTType;
+
+using concord::storage::sparse_merkle::BatchedInternalNode;
+using concord::storage::sparse_merkle::Hash;
+using concord::storage::sparse_merkle::Hasher;
+using concord::storage::sparse_merkle::IDBReader;
+using concord::storage::sparse_merkle::InternalChild;
+using concord::storage::sparse_merkle::InternalNodeKey;
+using concord::storage::sparse_merkle::LeafNode;
+using concord::storage::sparse_merkle::LeafKey;
+using concord::storage::sparse_merkle::LeafChild;
+using concord::storage::sparse_merkle::NibblePath;
+using concord::storage::sparse_merkle::Tree;
+
+namespace concord {
+namespace storage {
+namespace blockchain {
+
+namespace v2MerkleTree {
+
+namespace {
+
+// Specifies the key type. Used when serializing the stale node index.
+enum class KeySerializationType : std::uint8_t {
+  Internal,
+  Leaf,
+};
+
+// Specifies the child type. Used when serializing BatchedInternalNode objects.
+enum class ChildSerializationType : std::uint8_t { Leaf, Internal };
+
+template <typename E>
+constexpr auto toChar(E e) {
+  static_assert(std::is_enum_v<E>);
+  return static_cast<char>(e);
+}
+
+// Serialize integral types in big-endian (network) byte order so that lexicographical comparison works and we can get
+// away without a custom key comparator. Do not allow bools.
+template <typename T>
+std::string serializeImp(T v) {
+  static_assert(std::is_integral_v<T> && !std::is_same_v<T, bool>);
+
+  v = concordUtils::hostToNet(v);
+
+  const auto data = reinterpret_cast<const char *>(&v);
+  return std::string{data, sizeof(v)};
+}
+
+std::string serializeImp(EDBKeyType type) { return std::string{toChar(type)}; }
+
+std::string serializeImp(EKeyType type) { return std::string{toChar(EDBKeyType::Key), toChar(type)}; }
+
+std::string serializeImp(EBFTType type) { return std::string{toChar(EDBKeyType::BFT), toChar(type)}; }
+
+std::string serializeImp(KeySerializationType type) { return std::string{toChar(type)}; }
+
+// std::string serializeImp(ChildSerializationType type) { return std::string{toChar(type)}; }
+
+std::string serializeImp(const std::vector<std::uint8_t> &v) { return std::string{std::cbegin(v), std::cend(v)}; }
+
+std::string serializeImp(const NibblePath &path) {
+  return serializeImp(static_cast<std::uint8_t>(path.length())) + serializeImp(path.data());
```

Comment:
I think you are right.

My thinking was:

- having a simple interface that has an input and output is cleaner
- std::string has a small string optimisation (maybe around 20-24 bytes) and also a strategy for allocation that allocates more memory (e.g. twice the needed size)
- operator+ has overloads with rvalue references and we typically chain temporaries

Basically, based on above, it shouldn't allocate often (haven't measured it, though) and it can sometimes even avoid allocation.

I agree on the benchmarks and measurements, let's investigate that later.